storage/router/router.go

```diff
@@ -0,0 +1,139 @@
+package router
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/garyburd/redigo/redis"
+	logging "github.com/op/go-logging"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	log = logging.MustGetLogger("router")
+)
+
+/*
+Router defines a router engine. The router will try the pools in the same
+order defined by the table. And write to cache if wasn't retrieved from the
+cache pool
+*/
+type Router struct {
+	pools map[string]Pool
+
+	lookup []string
+	cache  []string
+}
+
+func (r *Router) get(key string) (string, []byte, error) {
+	for _, poolName := range r.lookup {
+		pool, ok := r.pools[poolName]
+		if !ok {
+			return "", nil, ErrPoolNotFound
+		}
+
+		data, err := pool.Get(key)
+		//only try next entry if entry is not found in this pool, or not routable
+		//otherwise return (nil, or other errors)
+		if err == ErrNotRoutable || err == redis.ErrNil {
+			continue
+		} else if err != nil {
+			log.Errorf("pool(%s, %s) : %s", poolName, key, err)
+			continue
+		}
+
+		return poolName, data, err
+	}
+
+	return "", nil, errors.Wrap(ErrNotRoutable, "no pools matches key")
+}
+
+/*
+updateCache will update keys that does not exist in local cache
+currently this is done sequentially, which means a GET for a block
+will not succeed unless the cache is updated. The errors of cache SET
+will only be logged thought and should not cause the GET operation to fail
+*/
+func (r *Router) updateCache(src, key string, data []byte) error {
```

Comment:
doing this sequentially will have a big impact on performance, can we do this async instead ?

Reply:
I know this can has a big impact on the system. I thought about doing it asynchronusly but I was too worried that we might end up with too many routines that tries to write blocks to redis which can cause pool exhaustion. Or if we have a limited number of workers (say 10) we can still hit a blocking limit again. We can never tell what are the best number of routines to commit to the cache.

I ended up deciding to do sequentially since this happens to the blocks that are NOT retrieved from cache.

If u think having a limited (say 100) workers to commit to cache is a better solution we can do it.

Reply:
I would just have a single gorountine to write to all the cache, so in  the `Get` method do `go r.updateCache(src, key, data)`

Regarding exhausting the redis pool, how is that a problem ? This code will run in goruntine, so you can even block when trying to get a connection from the redis pool it doesn't matter right ? you don't block any code path.